cmd/scollector/collectors/aws.go

```diff
@@ -0,0 +1,313 @@
+	descAWSEC2StatusCheck = "The EC2 Status Check, which includes both instance-level and system-level drill-down, gathered every 60 seconds."
+	descAWSELBLatency     = "The minimum, maximum and average latency as reported by the load balancer, gathered at a 60 second interval and averaged over five minutes."
+	descAWSELBHostCount   = "The number of instances in what the Elastic Load Balancer considers a healthy state, gathered every 60 seconds."
+)
+
+func AWS(accessKey, secretKey, region string) {
+	collectors = append(collectors, &IntervalCollector{
+		F: func() (opentsdb.MultiDataPoint, error) {
+			return c_aws(accessKey, secretKey, region)
+		},
+		Interval: 60 * time.Second,
+		name:     fmt.Sprintf("aws-%s", region),
+	})
+}
+
+func c_aws(accessKey, secretKey, region string) (opentsdb.MultiDataPoint, error) {
+	var md opentsdb.MultiDataPoint
+	creds := aws.Creds(accessKey, secretKey, "")
+	if creds == nil {
+		return nil, fmt.Errorf("unable to make creds")
+	}
+	ecc := ec2.New(creds, region, nil)
+	if ecc == nil {
+		return nil, fmt.Errorf("unable to login to EC2")
+	}
+	elb := elb.New(creds, region, nil)
+	if elb == nil {
+		return nil, fmt.Errorf("unable to login to ELB")
+	}
+	cw := cloudwatch.New(creds, region, nil)
+	if cw == nil {
+		return nil, fmt.Errorf("unable to login to CloudWatch")
+	}
+	instances, err := awsGetInstances(*ecc)
+	if err != nil {
+		slog.Info("No EC2 Instances found.")
+	}
+	loadBalancers, err := awsGetLoadBalancers(*elb)
+	if err != nil {
+		slog.Info("No ELB Load Balancecrs found.")
+	}
+	for _, loadBalancer := range loadBalancers {
+		awsGetELBLatency(*cw, &md, loadBalancer)
+		awsGetELBHostCounts(*cw, &md, loadBalancer)
+	}
+	for _, instance := range instances {
+		awsGetCPU(*cw, &md, instance)
+		awsGetNetwork(*cw, &md, instance)
+		awsGetDiskBytes(*cw, &md, instance)
+		awsGetDiskOps(*cw, &md, instance)
+		awsGetStatusChecks(*cw, &md, instance)
+	}
+	return md, nil
+}
+
+func awsGetInstances(ecc ec2.EC2) ([]ec2.Instance, error) {
+	instancelist := []ec2.Instance{}
+	resp, err := ecc.DescribeInstances(nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to describe EC2 Instances")
+	}
+	for _, reservation := range resp.Reservations {
+		for _, instance := range reservation.Instances {
+			instancelist = append(instancelist, instance)
+		}
+	}
+	return instancelist, nil
+}
+
+func awsGetLoadBalancers(lb elb.ELB) ([]elb.LoadBalancerDescription, error) {
+	lbList := []elb.LoadBalancerDescription{}
+	resp, err := lb.DescribeLoadBalancers(nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to describe ELB Balancers")
+	}
+	for _, loadBalancer := range resp.LoadBalancerDescriptions {
+		lbList = append(lbList, loadBalancer)
+	}
+	return lbList, nil
+}
+
+func awsGetCPU(cw cloudwatch.CloudWatch, md *opentsdb.MultiDataPoint, instance ec2.Instance) error {
+	search := cloudwatch.GetMetricStatisticsInput{
```

Comment:
@mjibson re: your quote about GetMetricStatisticsInput, it doesn't actually matter after thinking about it -- cloudwatch has already digested the data and I get X data points every Y minutes, so the cloudwatch query retrieves a predigested set of data.  I could theoretically initialize Start and the search struct in c_aws and pass that along to all the functions but we get back to that earlier conversation of "is it more readable that way?"  I'm willing to go either way on that.